[Issue-Resolver] Fix #33264 - RadioButtonGroup not working with Collection View

kubaflo opened the pull request:
# Fix #33264 - RadioButtonGroup not working with CollectionView

Fixes #33264

> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

## Summary

Fixed `RadioButtonGroup.SelectedValue` binding not working when RadioButtons are inside a CollectionView's ItemTemplate while RadioButtonGroup attached properties are set on a parent layout outside the CollectionView.

**Quick verification:**
- ✅ Tested on Android - Issue resolved, all tests passing
- ✅ Tested on iOS - Issue resolved (user verified)
- ✅ Edge cases tested
- ✅ UI tests added and passing

## Root Cause

RadioButtonGroupController only registered RadioButtons through `ChildAdded` events. When RadioButtons were dynamically created inside a CollectionView's ItemTemplate, they never triggered `ChildAdded` events on the parent layout where `RadioButtonGroup.GroupName` and `RadioButtonGroup.SelectedValue` were set.

Without registration, when a RadioButton was checked, `GetGroupController()` returned null, so `HandleRadioButtonGroupSelectionChanged()` was never called and the `SelectedValue` binding was never updated.

## Solution

Enhanced RadioButton to actively search up the visual tree for a matching RadioButtonGroup controller when checked, rather than relying solely on event-based registration:

**Key Changes:**

1. **`RadioButton.OnGroupSelectionChanged()`** - If no controller registered, searches ancestors for matching group and registers dynamically

2. **`RadioButton.FindControllerInAncestors()`** - Walks up visual tree to find layout with matching `RadioButtonGroup.GroupName`

3. **`RadioButtonGroupController.RegisterRadioButton()`** - Public method to register RadioButton with controller

4. **`RadioButtonGroup.GetController()`** - Exposes controller access for RadioButton

5. **`RadioButtonGroupController.SetSelectedValue()`** and **`SetGroupName()`** - Enhanced to proactively register discovered RadioButtons

**Files Changed:**
- `src/Controls/src/Core/RadioButton/RadioButton.cs` - Added ancestor search logic (+33 lines)
- `src/Controls/src/Core/RadioButton/RadioButtonGroupController.cs` - Added dynamic registration (+55 lines)
- `src/Controls/src/Core/RadioButton/RadioButtonGroup.cs` - Exposed controller access (+5 lines)

## Testing

**Before fix (Android):**
- Initial SelectedValue: '' 
- After tapping 'Choice 2', SelectedValue: '' ❌

**After fix (Android):**
- Initial SelectedValue: 'None' 
- After tapping 'Choice 2', SelectedValue: 'Choice 2' ✅ 
- After tapping 'Choice 3', SelectedValue: 'Choice 3' ✅
- Test Run Successful. Total tests: 1  Passed: 1

**After fix (iOS):**
- User verified - working correctly ✅

**Edge Cases Tested:**
- ✅ RadioButtons in CollectionView DataTemplate - Primary scenario, now working
- ✅ Multiple radio button selections - Sequential taps update binding correctly
- ✅ Binding updates in both directions - Property changes verified

**Platforms Tested:**
- ✅ Android - All tests passing
- ✅ iOS - User verified working

## Test Coverage

- ✅ Test page: `src/Controls/tests/TestCases.HostApp/Issues/Issue33264.xaml`
- ✅ Code-behind: `src/Controls/tests/TestCases.HostApp/Issues/Issue33264.xaml.cs`
- ✅ NUnit test: `src/Controls/tests/TestCases.Shared.Tests/Tests/Issues/Issue33264.cs`
- ✅ Tests exact scenario from issue report

## Breaking Changes

None - the fix adds proactive discovery without changing existing behavior for scenarios that already work.

kubaflo committed:
Handle descendant add/remove in RadioButtonGroupController

Added event handlers for DescendantAdded and DescendantRemoved to ensure radio buttons added or removed as descendants are properly managed by the group controller. This improves dynamic handling of radio button groups when the visual tree changes.

kubaflo committed:
[Issue-Resolver] Fix #33264 - RadioButtonGroup not working with CollectionView

kubaflo committed:
Revert "[Issue-Resolver] Fix #33264 - RadioButtonGroup not working with CollectionView"

This reverts commit 8a3d4cad8edf6d5425051f48f516c19ce08a1f07.

kubaflo committed:
Update Issue33264.cs

kubaflo committed:
Add a unit test

kubaflo commented:
Answer to https://github.com/dotnet/maui/pull/33268#pullrequestreview-3619357490
@PureWeen when I insisted more on creating a unit tests it did create them :)

Also I asked it to come up with UI tests without Task.Delays

 I've successfully added 2 unit tests that validate the RadioButtonGroup + CollectionView fix without requiring UI rendering or Appium.

   Unit Tests Added

   Location: src/Controls/tests/Core.UnitTests/RadioButtonTests.cs

   Test 1: RadioButtonGroupWorksWithDynamicallyAddedDescendants

   Tests that RadioButtons added as descendants (not direct children) get properly registered with the RadioButtonGroup controller and update the SelectedValue binding when checked.

   What it tests:

     - RadioButtons nested inside containers receive the group name from ancestors ✅
     - Checking a nested RadioButton updates the parent's SelectedValue ✅
     - Group selection enforcement works (only one checked at a time) ✅

   Test 2: RadioButtonGroupSelectedValueBindingWorksWithNestedDescendants

   Tests that setting SelectedValue on the group finds and checks the correct descendant RadioButton, even when deeply nested.

   What it tests:

     - Setting SelectedValue selects the correct nested RadioButton ✅
     - Changing SelectedValue updates selection correctly ✅
     - Bidirectional binding works (group ↔️ RadioButton) ✅

   Test Results

     ✅ Passed!  - Failed: 0, Passed: 2, Total: 2
     ✅ All RadioButton tests: Failed: 0, Passed: 17, Total: 17

   How They Work

   The tests simulate CollectionView behavior by:

     - Creating a parent layout with RadioButtonGroup.GroupName and RadioButtonGroup.SelectedValue
     - Adding a nested container (simulates CollectionView item container)
     - Adding RadioButtons to the nested container (triggers DescendantAdded events)
     - Verifying the fix's DescendantAdded/DescendantRemoved event handlers work correctly

   Why These Are True Unit Tests

   ✅ No UI rendering required - Tests pure control logic ✅ No Appium/platform code - Tests run in-process with xUnit ✅ Fast execution - 22ms vs. ~60+ seconds for UI tests ✅ Test the
   actual fix - Exercise the new DescendantAdded/DescendantRemoved handlers ✅ Platform-agnostic - Test logic that applies to all platforms

   Comparison

   ┌────────────────┬─────────────────────────────────────┬───────────────────────────────┐
   │ Aspect         │ UI Test                             │ Unit Test                     │
   ├────────────────┼─────────────────────────────────────┼───────────────────────────────┤
   │ Execution time │ ~60-90 seconds                      │ 22ms (2700x faster)           │
   ├────────────────┼─────────────────────────────────────┼───────────────────────────────┤
   │ Requirements   │ Emulator/Simulator, Appium          │ None                          │
   ├────────────────┼─────────────────────────────────────┼───────────────────────────────┤
   │ What it tests  │ Full rendering + binding + handlers │ Control logic + events        │
   ├────────────────┼─────────────────────────────────────┼───────────────────────────────┤
   │ When it runs   │ CI (platform-specific)              │ Every build (all platforms)   │
   ├────────────────┼─────────────────────────────────────┼───────────────────────────────┤
   │ Coverage       │ End-to-end validation               │ Core functionality validation │
   └────────────────┴─────────────────────────────────────┴───────────────────────────────┘

   Both test types are valuable:

     - UI tests validate the complete scenario with real rendering
     - Unit tests validate the core fix logic quickly and reliably

   The UI test remains important for full integration validation, but the unit tests provide fast feedback during development and catch regressions early.

PureWeen reviewed (changes requested):
UItests are much slower than unit tests so whenever we can avoid adding a UITest that's a win. 

Is it alright if we remove the UITests from this one?

kubaflo committed:
Remove test and sample for Issue33264

Deleted the sample page, view model, and UI test related to Issue33264 (RadioButtonGroup with CollectionView). This cleanup may indicate the issue is resolved, obsolete, or the test is no longer needed.

PureWeen commented on src/Controls/src/Core/RadioButton/RadioButtonGroupController.cs:
```diff
@@ -24,6 +24,8 @@ public RadioButtonGroupController(Maui.ILayout layout)
 			_layout = (Element)layout;
 			_layout.ChildAdded += ChildAdded;
 			_layout.ChildRemoved += ChildRemoved;
+			_layout.DescendantAdded += DescendantAdded;
```
Is the childAdded and ChildRemoved now redundant? can we get rid of those?

kubaflo committed:
Remove redundant ChildAdded and ChildRemoved event handlers

Eliminated the ChildAdded and ChildRemoved event subscriptions and their corresponding handler methods from RadioButtonGroupController, as these are no longer needed. The controller now relies solely on DescendantAdded and DescendantRemoved for managing radio button group membership.
